### ISSUE TITLE
Windows start lagging/stuttering after a few minutes unless I regularly switch workspaces.

### ISSUE THREAD

**Issue by starcraft66** [i3/i3#4233]
## I'm submitting a…

<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

After about 10 minutes of working without switching workspaces, a strange stuttering phenomenon starts to happen and the only way to make it go away is to switch workspaces back and forth but the problem will always keep persisting as time goes on.

I made a YouTube video describing the problem here: https://www.youtube.com/watch?v=SJrVwb6lbV4
I recorded it on my phone in case I couldn't properly capture the problem using a screen recorder.

## Expected Behavior

I expect there to be no stuttering and that scrolling/gaming always be smooth.

## Reproduction Instructions
1. Start i3, go do some work without switching workspaces like a long browsing session or playing a game.
2. Wait a while, not exactly sure how long but ~10 mins.
3. There will be frequent stuttering/frame skipping.

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
Binary i3 version:  4.18.2 (2020-07-26) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.18.2 (2020-07-26) (pid 28055)bort…)
Loaded i3 config: /home/tristan/.config/i3/config (Last modified: Wed 31 Dec 1969 07:00:01 PM EST, 1603862866 seconds ago)

The i3 binary you just called: /nix/store/lilap57kf1nc852ysycqllvrqkd631g6-i3-4.18.2/bin/i3
The i3 binary you are running: /nix/store/r57v05wl5yzsy8hfvzpf8ylfsv0gy3sf-i3-gaps-4.18.2/bin/i3
</pre>

<details><summary>Config file</summary><pre>
font pango:pango:MesloLGS Nerd Font Mono 10
floating_modifier Mod4
new_window normal 2
new_float normal 2
hide_edge_borders none
force_focus_wrapping no
focus_follows_mouse yes
focus_on_window_activation smart
mouse_warping output
workspace_layout default
workspace_auto_back_and_forth yes

client.focused #81a2be #81a2be #1d1f21 #282a2e #81a2be
client.focused_inactive #373b41 #373b41 #969896 #292d2e #373b41
client.unfocused #282a2e #282a2e #999999 #292d2e #282a2e
client.urgent #2f343a #cc6666 #a3685a #cc6666 #2f343a
client.placeholder #000000 #0c0c0c #ffffff #000000 #0c0c0c
client.background #ffffff

bindsym Mod1+Tab exec /nix/store/iyp5i394clnzz67qilf82xvvfbd99gzl-rofi-1.5.4/bin/rofi -show window
bindsym Mod4+0 workspace 10
bindsym Mod4+1 workspace 1
bindsym Mod4+2 workspace 2
bindsym Mod4+3 workspace 3
bindsym Mod4+4 workspace 4
bindsym Mod4+5 workspace 5
bindsym Mod4+6 workspace 6
bindsym Mod4+7 workspace 7
bindsym Mod4+8 workspace 8
bindsym Mod4+9 workspace 9
bindsym Mod4+Down focus down
bindsym Mod4+Left focus left
bindsym Mod4+Print exec flameshot gui
bindsym Mod4+Return exec /nix/store/r52s0b6d935iaj78dki36sig9w79hkwr-alacritty-0.5.0/bin/alacritty
bindsym Mod4+Right focus right
bindsym Mod4+Shift+0 move container to workspace 10
bindsym Mod4+Shift+1 move container to workspace 1
bindsym Mod4+Shift+2 move container to workspace 2
bindsym Mod4+Shift+3 move container to workspace 3
bindsym Mod4+Shift+4 move container to workspace 4
bindsym Mod4+Shift+5 move container to workspace 5
bindsym Mod4+Shift+6 move container to workspace 6
bindsym Mod4+Shift+7 move container to workspace 7
bindsym Mod4+Shift+8 move container to workspace 8
bindsym Mod4+Shift+9 move container to workspace 9
bindsym Mod4+Shift+Down move down
bindsym Mod4+Shift+Left move left
bindsym Mod4+Shift+Right move right
bindsym Mod4+Shift+Up move up
bindsym Mod4+Shift+bar split h
bindsym Mod4+Shift+c reload
bindsym Mod4+Shift+e exec "i3-nagbar -t warning -m 'You pressed the exit shortcut. Do you really want to exit i3? This will end your X session.' -b 'Yes, exit i3' 'i3-msg exit'"
bindsym Mod4+Shift+f floating toggle
bindsym Mod4+Shift+grave move scratchpad
bindsym Mod4+Shift+h move left
bindsym Mod4+Shift+j move down
bindsym Mod4+Shift+k move up
bindsym Mod4+Shift+l move right
bindsym Mod4+Shift+q kill
bindsym Mod4+Shift+r restart
bindsym Mod4+Shift+s layout stacking
bindsym Mod4+Shift+t layout tabbed
bindsym Mod4+Space focus mode_toggle
bindsym Mod4+Up focus up
bindsym Mod4+a exec /nix/store/iyp5i394clnzz67qilf82xvvfbd99gzl-rofi-1.5.4/bin/rofi -show run
bindsym Mod4+e exec thunar
bindsym Mod4+f fullscreen
bindsym Mod4+grave scratchpad show
bindsym Mod4+h focus left
bindsym Mod4+j focus down
bindsym Mod4+k focus up
bindsym Mod4+l focus right
bindsym Mod4+minus split v
bindsym Mod4+p exec /nix/store/dk0skb4mm4z309xpfk33akq5x542j0la-rofi-pass-2.0.2/bin/rofi-pass
bindsym Mod4+q exec dmenu_run
bindsym Mod4+r mode resize
bindsym Mod4+w exec chromium
bindsym XF86AudioLowerVolume exec /nix/store/l9rj3f2znh4qplbikdanw1j1c63lxd9x-pulseaudio-13.0/bin/pactl set-sink-volume 0 -5%
bindsym XF86AudioMute exec /nix/store/l9rj3f2znh4qplbikdanw1j1c63lxd9x-pulseaudio-13.0/bin/pactl set-sink-mute 0 toggle
bindsym XF86AudioNext exec /nix/store/2zgliqcj65dnnph9ik3yyvy1a1ia5ra1-playerctl-2.2.1/bin/playerctl next
bindsym XF86AudioPause exec /nix/store/2zgliqcj65dnnph9ik3yyvy1a1ia5ra1-playerctl-2.2.1/bin/playerctl pause
bindsym XF86AudioPlay exec /nix/store/2zgliqcj65dnnph9ik3yyvy1a1ia5ra1-playerctl-2.2.1/bin/playerctl play
bindsym XF86AudioPrev exec /nix/store/2zgliqcj65dnnph9ik3yyvy1a1ia5ra1-playerctl-2.2.1/bin/playerctl previous
bindsym XF86AudioRaiseVolume exec /nix/store/l9rj3f2znh4qplbikdanw1j1c63lxd9x-pulseaudio-13.0/bin/pactl set-sink-volume 0 +5%
bindsym XF86MonBrightnessDown exec /nix/store/8sh8vr96wkhx7zrkrfs9xc8ppcdhgpd2-xbacklight-1.2.3/bin/xbacklight -dec 5
bindsym XF86MonBrightnessUp exec /nix/store/8sh8vr96wkhx7zrkrfs9xc8ppcdhgpd2-xbacklight-1.2.3/bin/xbacklight -inc 5

mode "resize" {
bindsym Down resize grow height 10 px or 10 ppt
bindsym Escape mode default
bindsym Left resize shrink width 10 px or 10 ppt
bindsym Return mode default
bindsym Right resize grow width 10 px or 10 ppt
bindsym Up resize shrink height 10 px or 10 ppt
bindsym h resize shrink width 10 px or 10 ppt
bindsym j resize shrink height 10 px or 10 ppt
bindsym k resize grow height 10 px or 10 ppt
bindsym l resize grow width 10 px or 10 ppt
}

bar {
  
  font pango:pango:MesloLGS Nerd Font Mono 10
  mode dock
  hidden_state hide
  position top
  status_command /nix/store/ybwnylvwybb2a8w0bs0x0fp2r5lf332i-i3status-2.13/bin/i3status
  i3bar_command /nix/store/lilap57kf1nc852ysycqllvrqkd631g6-i3-4.18.2/bin/i3bar
  workspace_buttons yes
  strip_workspace_numbers no
  tray_output DP-4
  colors {
    background #282a2e
    statusline #282a2e
    separator #515151
    focused_workspace #81a2be #81a2be #1d1f21
    active_workspace #333333 #333333 #a3685a
    inactive_workspace #282a2e #282a2e #999999
    urgent_workspace #cc6666 #cc6666 #a3685a
    binding_mode #2f343a #900000 #ffffff
  }
  
}

exec --no-startup-id pkill picom; /nix/store/8addp0br1psrw8nry1yzvw6pi39dygh2-picom-8/bin/picom --backend xrender --xrender-sync-fence --no-vsync

exec --no-startup-id pkill ckb-next; /nix/store/nv4dkwpxgrsphsdvgh22jgnbmhswb59h-ckb-next-0.4.2/bin/ckb-next --background

exec --no-startup-id pkill xsecurelock; /nix/store/9c4vxgjacfhhn4mh3bca7rhpn9jgsl3z-xss-lock-git-2018-05-31/bin/xss-lock /nix/store/w9wc0d31p4z93cbgxijws03j5s2c4gyf-coreutils-8.31/bin/env XSECURELOCK_PASSWORD_PROMPT=time_hex XSECURELOCK_NO_COMPOSITE=1 XSECURELOCK_BLANK_DPMS_STATE=off XSECURELOCK_BLANK_TIMEOUT=30 /nix/store/y0yn2ssmv7n4svxrkvang8q55h6a1na3-xsecurelock-1.7.0/bin/xsecurelock
</pre>
</details>

<pre>
Logfile URL:
</pre>

## Other details
I originally thought this was an issue with my compositor, Picom, but it turns out I can reproduce the issue even without a compositor running.

Video of the problem in action:

https://www.youtube.com/watch?v=SJrVwb6lbV4

There are timestamps in the video description.

Note that I originally found this bug when I was running the i3-gaps fork so I switched to plain i3 and it still seems to happen. The Xorg server process usually has higher cpu usage than normal when this issue is occurring, sometimes hitting 100% of a core when the lagging ins unbearable.

<pre>
- Linux Distribution & Version: NixOS 20.09
- Are you using a compositor (e.g., xcompmgr or compton): I am using Picom
</pre>

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow https://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by orestisfl** [comment]
Does it happen without picom?

**Comment by stapelberg** [comment]
> > Does it happen without picom?

Yes, as the report states:

> I originally thought this was an issue with my compositor, Picom, but it turns out I can reproduce the issue even without a compositor running.

---

Can you check with a different window manager that this problem is unrelated to your GPU etc.?

**Comment by starcraft66** [comment]
> Can you check with a different window manager that this problem is unrelated to your GPU etc.?

The only other DE/WM I have installed is KDE Plasma and I can't reproduce the problem with it. I usually switch to Plasma to play video games so that I can avoid the periodic stuttering that builds up.

**Comment by stapelberg** [comment]
Can you install awesome or some other window manager? We really need to narrow down this problem more first.

**Comment by stapelberg** [comment]
Oh, I just remembered: your symptoms match overly-aggressive powersaving measures of modern GPUs.

Which GPU do you have?

Have you tried setting it to “maximum performance” and see if that solves the issue?

**Comment by starcraft66** [comment]
![image](https://user-images.githubusercontent.com/1858154/97468137-52450b00-191b-11eb-9bf0-cb67849dbf0d.png)

I have a GTX 1080 Ti running the NVIDIA Proprietary driver.

I see in the control panel that the power mode is on adaptive. I will set it to max performance and check if the issue still reproduces.

**Comment by starcraft66** [comment]
Update: Setting the card to Max Performance doesn't change anything. I watched a YouTube video in firefox for about 15 minutes and just a few minutes in the video and the rest of the system was already stuttering every second with the X server using a lot more CPU than it should.

**Comment by orestisfl** [comment]
> Can you install awesome or some other window manager? We really need to narrow down this problem more first.

@starcraft66 please let us know if you have the chance to try with some other wm. openbox or awesomewm are some widely-available choices and using them while browsing shouldn't affect your workflow.

**Comment by starcraft66** [comment]
@orestisfl Sorry for the long delay, I was doing some testing on my end. It looks like upgrading to the Nvidia 455 driver fixed the stuttering issue. I started using i3 again around the same time I updated to the Nvidia 450 driver and it looks like that driver has a serious issue. I've been using my computer with i3 and the Nvidia 455 driver for a week now and I have not been able to reproduce the stuttering.
